File: src/samovar/sparsedossa2.R
#!/usr/bin/env Rscript
# SparseDOSSA2 driver for SamovaR table generation and CV scoring.
# Not an R package; only calls exported SparseDOSSA2 / future functions.

args <- commandArgs(trailingOnly = TRUE)

`%||%` <- function(a, b) if (is.null(a) || (length(a) == 1 && is.na(a))) b else a

parse_args <- function(argv) {
  out <- list(
    command = "simulate",
    input = NULL,
    output = NULL,
    template = "Stool",
    mode = "fit",
    n_sample = 10L,
    n_feature = NULL,
    new_features = NA,
    seed = 42L,
    workers = 1L,
    verbose = FALSE,
    cv_folds = 5L,
    lambdas = NULL,
    fit_lambda = NULL,
    maxit = NULL,
    max_eval = NULL,
    prec_bits = NULL
  )
  if (length(argv) >= 1 && !startsWith(argv[[1]], "-")) {
    out$command <- argv[[1]]
    argv <- argv[-1]
  }
  i <- 1
  while (i <= length(argv)) {
    key <- argv[[i]]
    take <- function() {
      if (i + 1 > length(argv)) stop("Missing value for ", key)
      argv[[i + 1]]
    }
    if (key %in% c("--input", "-i")) {
      out$input <- take(); i <- i + 2
    } else if (key %in% c("--output", "-o")) {
      out$output <- take(); i <- i + 2
    } else if (key == "--template") {
      out$template <- take(); i <- i + 2
    } else if (key == "--mode") {
      out$mode <- take(); i <- i + 2
    } else if (key %in% c("--n-sample", "--n_sample", "--N")) {
      out$n_sample <- as.integer(take()); i <- i + 2
    } else if (key %in% c("--n-feature", "--n_feature", "--n-features")) {
      out$n_feature <- as.integer(take()); i <- i + 2
    } else if (key %in% c("--new-features", "--new_features")) {
      nxt <- if (i + 1 <= length(argv) && !startsWith(argv[[i + 1]], "-")) {
        val <- take(); i <- i + 2; val
      } else {
        i <- i + 1; "TRUE"
      }
      out$new_features <- toupper(as.character(nxt)) %in% c("1", "TRUE", "YES", "T")
    } else if (key == "--seed") {
      out$seed <- as.integer(take()); i <- i + 2
    } else if (key %in% c("--workers", "--parallel", "--cores")) {
      out$workers <- as.integer(take()); i <- i + 2
    } else if (key %in% c("--verbose", "-v")) {
      if (i + 1 <= length(argv) && !startsWith(argv[[i + 1]], "-")) {
        out$verbose <- toupper(take()) %in% c("1", "TRUE", "YES", "T"); i <- i + 2
      } else {
        out$verbose <- TRUE; i <- i + 1
      }
    } else if (key %in% c("--cv-folds", "--K", "-K")) {
      out$cv_folds <- as.integer(take()); i <- i + 2
    } else if (key %in% c("--lambdas", "--lambda-grid")) {
      out$lambdas <- take(); i <- i + 2
    } else if (key %in% c("--lambda", "--fit-lambda")) {
      out$fit_lambda <- as.numeric(take()); i <- i + 2
    } else if (key %in% c("--maxit", "--max-it")) {
      out$maxit <- as.integer(take()); i <- i + 2
    } else if (key %in% c("--max-eval", "--max_eval")) {
      out$max_eval <- as.integer(take()); i <- i + 2
    } else if (key %in% c("--prec-bits", "--precBits", "--prec_bits")) {
      out$prec_bits <- as.integer(take()); i <- i + 2
    } else if (key == "--fit") {
      out$mode <- "fit"; i <- i + 1
    } else {
      i <- i + 1
    }
  }
  out
}

limit_blas_threads <- function() {
  for (var in c("OMP_NUM_THREADS", "OPENBLAS_NUM_THREADS", "MKL_NUM_THREADS", "NUMEXPR_NUM_THREADS")) {
    if (!nzchar(Sys.getenv(var, unset = ""))) {
      Sys.setenv(var, "1")
    }
  }
}

make_control <- function(opts) {
  control <- list(verbose = isTRUE(opts$verbose))
  if (!is.null(opts$maxit) && is.finite(opts$maxit) && opts$maxit >= 1L) {
    control$maxit <- as.integer(opts$maxit)
  }
  numint <- list()
  if (!is.null(opts$max_eval) && is.finite(opts$max_eval) && opts$max_eval >= 1L) {
    numint$max_eval <- as.integer(opts$max_eval)
  }
  if (!is.null(opts$prec_bits) && is.finite(opts$prec_bits) && opts$prec_bits >= 1L) {
    numint$precBits <- as.integer(opts$prec_bits)
  }
  if (length(numint)) {
    control$control_numint <- numint
  }
  control
}

setup_parallel <- function(workers, cv = FALSE) {
  workers <- as.integer(workers %||% 1L)
  if (!requireNamespace("future", quietly = TRUE)) {
    return(invisible(NULL))
  }
  # Wiki uses future::plan(sequential(), ..., multisession()). On future 1.70
  # sequential() must not be called; nested CV uses a strategy list + tweak().
  if (is.na(workers) || workers <= 1L) {
    future::plan(future::sequential)
    return(invisible(NULL))
  }
  ms <- future::tweak(future::multisession, workers = workers)
  if (cv) {
    future::plan(list(future::sequential, future::sequential, ms))
  } else {
    future::plan(ms)
  }
  invisible(NULL)
}

read_feature_matrix <- function(path) {
  raw <- utils::read.csv(path, row.names = 1, check.names = FALSE)
  mat <- as.matrix(raw)
  storage.mode(mat) <- "numeric"
  if (nrow(mat) < ncol(mat)) {
    mat <- t(mat)
  }
  mat[!is.finite(mat)] <- 0
  mat
}

write_json <- function(path, obj) {
  txt <- paste0("{", paste(vapply(names(obj), function(k) {
    val <- obj[[k]]
    if (is.null(val) || (length(val) == 1 && is.na(val))) {
      sprintf('"%s": null', k)
    } else if (is.logical(val) && length(val) == 1) {
      sprintf('"%s": %s', k, if (isTRUE(val)) "true" else "false")
    } else if (is.numeric(val) && length(val) == 1) {
      sprintf('"%s": %s', k, format(val, digits = 16, scientific = TRUE))
    } else {
      sprintf('"%s": %s', k, json_quote(as.character(val)[[1]]))
    }
  }, character(1)), collapse = ", "), "}\n")
  writeLines(txt, path, sep = "")
}

json_quote <- function(x) {
  x <- gsub("\\", "\\\\", x, fixed = TRUE)
  x <- gsub("\"", "\\\"", x, fixed = TRUE)
  x <- gsub("\n", "\\n", x, fixed = TRUE)
  paste0("\"", x, "\"")
}

extract_cv_score <- function(fitted) {
  gof <- fitted$cv_goodness_of_fit
  if (!is.null(gof)) {
    if (is.numeric(gof) && length(gof) >= 1) {
      return(mean(as.numeric(gof), na.rm = TRUE))
    }
    if (is.list(gof)) {
      for (key in c("mean", "score", "value", "cv_goodness_of_fit")) {
        if (!is.null(gof[[key]]) && is.numeric(gof[[key]])) {
          return(mean(as.numeric(gof[[key]]), na.rm = TRUE))
        }
      }
    }
  }
  ll <- fitted$EM_fit$logLik_CV
  if (!is.null(ll)) {
    return(mean(as.numeric(ll), na.rm = TRUE))
  }
  NA_real_
}

opts <- parse_args(args)
limit_blas_threads()
if (is.null(opts$output)) stop("--output is required")
if (!requireNamespace("SparseDOSSA2", quietly = TRUE)) {
  stop("SparseDOSSA2 is not installed. Run ./install.sh SparseDOSSA2")
}
suppressPackageStartupMessages(library(SparseDOSSA2))

if (opts$command %in% c("fitcv", "cv", "score")) {
  if (is.null(opts$input)) stop("--input is required for fitcv")
  mat <- read_feature_matrix(opts$input)
  n_samp <- ncol(mat)
  k <- max(2L, min(as.integer(opts$cv_folds %||% 5L), n_samp))
  setup_parallel(opts$workers, cv = TRUE)
  if (!is.null(opts$seed) && !is.na(opts$seed)) set.seed(opts$seed)
  control <- make_control(opts)
  lambda_vec <- NULL
  if (!is.null(opts$lambdas) && nzchar(opts$lambdas)) {
    lambda_vec <- as.numeric(strsplit(opts$lambdas, ",", fixed = TRUE)[[1]])
    lambda_vec <- lambda_vec[is.finite(lambda_vec)]
  }
  fit_args <- list(data = mat, K = k, control = control)
  if (length(lambda_vec)) fit_args$lambdas <- lambda_vec
  fitted <- do.call(fitCV_SparseDOSSA2, fit_args)
  score <- extract_cv_score(fitted)
  write_json(opts$output, list(
    cv_goodness_of_fit = score,
    K = k,
    n_feature = nrow(mat),
    n_sample = ncol(mat)
  ))
  quit(save = "no", status = 0)
}

setup_parallel(opts$workers, cv = FALSE)
if (!is.null(opts$seed) && !is.na(opts$seed)) set.seed(opts$seed)

new_features <- opts$new_features
if (identical(opts$mode, "fit") || identical(tolower(opts$template), "fit")) {
  if (is.null(opts$input)) stop("--input is required for fit mode")
  mat <- read_feature_matrix(opts$input)
  if (is.na(new_features)) new_features <- FALSE
  n_feature <- opts$n_feature %||% nrow(mat)
  control <- make_control(opts)
  lambda <- 1
  if (!is.null(opts$fit_lambda) && is.finite(opts$fit_lambda)) {
    lambda <- opts$fit_lambda
  }
  fitted <- fit_SparseDOSSA2(data = mat, lambda = lambda, control = control)
  sim <- SparseDOSSA2(
    template = fitted,
    n_sample = as.integer(opts$n_sample),
    n_feature = as.integer(n_feature),
    new_features = isTRUE(new_features),
    verbose = isTRUE(opts$verbose)
  )
} else {
  if (is.na(new_features)) new_features <- TRUE
  n_feature <- opts$n_feature %||% 100L
  sim <- SparseDOSSA2(
    template = as.character(opts$template),
    n_sample = as.integer(opts$n_sample),
    n_feature = as.integer(n_feature),
    new_features = isTRUE(new_features),
    verbose = isTRUE(opts$verbose)
  )
}

counts <- sim$simulated_data
utils::write.csv(counts, opts$output, row.names = TRUE)
quit(save = "no", status = 0)

File: src/samovar/sparsedossa2.py
"""SparseDOSSA2 abundance-table generator and CV table scorer.

Optional R package: https://github.com/biobakery/SparseDOSSA2
Wiki: https://github.com/biobakery/biobakery/wiki/SparseDOSSA2

Install with ``./install.sh SparseDOSSA2``. Parallelism uses the R ``future``
plan from the wiki (``multisession``; CV: sequential/sequential/multisession).
"""

from __future__ import annotations

import itertools
import json
import os
import shutil
import subprocess
import tempfile
from pathlib import Path
from typing import Any, Dict, Iterable, List, Optional, Sequence, Tuple, Union

import pandas as pd

from samovar.parse_annotators import Annotation
from samovar.paths import load_config, repo_root
from samovar.regenerate import (
    SPARSEDOSSA2_MODE_ALIASES,
    _abundance_table_from_matrix,
    _annotator_name,
    _apply_abundance_scale,
    _count_matrix,
    _filter_taxa,
    _n_samples_or_observed,
    _taxid_columns,
    coerce_seed,
)
from samovar.table_regenerators import apply_extra_flags

PathLike = Union[str, Path]

TEMPLATES = ("Stool", "Vaginal", "IBD")
MODE_ALIASES = SPARSEDOSSA2_MODE_ALIASES
_JOB_SEQ = itertools.count(1)


def _job_tag(prefix: str) -> str:
    return f"{prefix}_{os.getpid()}_{next(_JOB_SEQ)}"


class MissingSparseDOSSA2Error(RuntimeError):
    """R package SparseDOSSA2 is not installed."""


def canonicalize_sparsedossa2_mode(name: Optional[str]) -> Optional[str]:
    key = str(name or "").strip().lower().replace("_", "-")
    return MODE_ALIASES.get(key)


def template_for_mode(mode: str) -> Optional[str]:
    canon = canonicalize_sparsedossa2_mode(mode) or str(mode)
    if canon.endswith("-stool"):
        return "Stool"
    if canon.endswith("-vaginal"):
        return "Vaginal"
    if canon.endswith("-ibd"):
        return "IBD"
    if canon.endswith("-fit") or canon == "sparsedossa2":
        return None
    return None


def rscript_executable() -> str:
    cfg = load_config()
    try:
        from samovar.main_config import iter_tools, tool_path

        spec = iter_tools(cfg).get("Rscript")
        stored = tool_path(spec, "Rscript") if spec else ""
        if stored and Path(stored).expanduser().is_file():
            return str(Path(stored).expanduser())
    except Exception:
        pass
    for key in ("rscript_path", "Rscript"):
        raw = str(cfg.get(key) or "").strip()
        if raw and Path(raw).expanduser().is_file():
            return str(Path(raw).expanduser())
    r_path = str(cfg.get("r_path") or "").strip()
    if r_path:
        candidate = Path(r_path).expanduser()
        if candidate.name in {"R", "R.exe"}:
            sibling = candidate.with_name("Rscript")
            if sibling.is_file():
                return str(sibling)
        rs = candidate.parent / "Rscript"
        if rs.is_file():
            return str(rs)
    found = shutil.which("Rscript")
    if found:
        return found
    raise MissingSparseDOSSA2Error(
        "Rscript is not on PATH. Install R, then ./install.sh SparseDOSSA2."
    )


def driver_script() -> Path:
    env = os.environ.get("SAMOVAR_SPARSEDOSSA2_R", "").strip()
    if env:
        path = Path(env).expanduser()
        if path.is_file():
            return path
    cfg = load_config()
    raw = str(cfg.get("sparsedossa2_r") or "").strip()
    if raw:
        path = Path(raw).expanduser()
        if path.is_file():
            return path
    try:
        from samovar.main_config import iter_tools, tool_path

        spec = iter_tools(cfg).get("sparsedossa2.R")
        stored = tool_path(spec, "sparsedossa2.R") if spec else ""
        if stored and Path(stored).expanduser().is_file():
            return Path(stored).expanduser()
    except Exception:
        pass
    packaged = Path(__file__).with_name("sparsedossa2.R")
    if packaged.is_file():
        return packaged
    fallback = repo_root() / "src" / "samovar" / "sparsedossa2.R"
    if fallback.is_file():
        return fallback
    raise MissingSparseDOSSA2Error("SparseDOSSA2 R driver sparsedossa2.R is missing.")


def sparsedossa2_available() -> bool:
    try:
        rscript = rscript_executable()
    except MissingSparseDOSSA2Error:
        return False
    try:
        proc = subprocess.run(
            [
                rscript,
                "--vanilla",
                "-e",
                'cat(if (requireNamespace("SparseDOSSA2", quietly=TRUE)) "TRUE" else "FALSE")',
            ],
            capture_output=True,
            text=True,
            timeout=60,
        )
    except (OSError, subprocess.SubprocessError):
        return False
    return "TRUE" in (proc.stdout or "")


def require_sparsedossa2() -> None:
    if not sparsedossa2_available():
        raise MissingSparseDOSSA2Error(
            "SparseDOSSA2 is not installed. Run ./install.sh SparseDOSSA2 "
            "(https://github.com/biobakery/SparseDOSSA2)."
        )


def _workers(config: Dict[str, Any]) -> int:
    for key in ("sparsedossa2_workers", "workers", "parallel", "cores"):
        raw = config.get(key)
        if raw in (None, "", False):
            continue
        try:
            return max(1, int(raw))
        except (TypeError, ValueError):
            continue
    env = os.environ.get("SAMOVAR_SPARSEDOSSA2_WORKERS") or os.environ.get(
        "SAMOVAR_WORKERS"
    )
    if env:
        try:
            return max(1, int(env))
        except ValueError:
            pass
    return 1


def _r_env() -> Dict[str, str]:
    env = os.environ.copy()
    for key in (
        "OMP_NUM_THREADS",
        "OPENBLAS_NUM_THREADS",
        "MKL_NUM_THREADS",
        "NUMEXPR_NUM_THREADS",
    ):
        env.setdefault(key, "1")
    return env


def _append_em_controls(argv: List[str], cfg: Dict[str, Any]) -> None:
    if cfg.get("maxit") not in (None, ""):
        argv.extend(["--maxit", str(int(cfg["maxit"]))])
    if cfg.get("max_eval") not in (None, ""):
        argv.extend(["--max-eval", str(int(cfg["max_eval"]))])
    if cfg.get("prec_bits") not in (None, ""):
        argv.extend(["--prec-bits", str(int(cfg["prec_bits"]))])


def _job_workdir(cfg: Optional[Dict[str, Any]] = None) -> Path:
    cfg = cfg or {}
    raw = (
        cfg.get("sd2_workdir")
        or cfg.get("output_dir")
        or os.environ.get("SAMOVAR_SD2_WORKDIR")
        or ""
    )
    if raw:
        path = Path(str(raw)).expanduser() / ".sd2_jobs"
        path.mkdir(parents=True, exist_ok=True)
        return path
    return Path(tempfile.mkdtemp(prefix="samovar_sd2_"))


def run_sparsedossa2_r(
    argv: Sequence[str],
    *,
    timeout: Optional[int] = None,
    config: Optional[Dict[str, Any]] = None,
) -> str:
    require_sparsedossa2()
    _ = config
    cmd = [rscript_executable(), "--vanilla", str(driver_script()), *[str(x) for x in argv]]
    try:
        proc = subprocess.run(
            cmd,
            capture_output=True,
            text=True,
            timeout=timeout,
            env=_r_env(),
        )
    except subprocess.TimeoutExpired as exc:
        raise RuntimeError(
            f"SparseDOSSA2 R driver timed out after {timeout}s"
        ) from exc
    if proc.returncode != 0:
        err = (proc.stderr or proc.stdout or "").strip()
        raise RuntimeError(
            "SparseDOSSA2 R driver failed:\n" + err[-4000:]
        )
    return proc.stdout or ""


def _write_feature_matrix(matrix: pd.DataFrame, dest: Path) -> None:
    dest.parent.mkdir(parents=True, exist_ok=True)
    work = matrix.copy()
    work.index = work.index.astype(str)
    work.to_csv(dest)


def _read_simulated_csv(path: Path) -> pd.DataFrame:
    frame = pd.read_csv(path, index_col=0)
    frame.index = frame.index.astype(str)
    return frame.apply(pd.to_numeric, errors="coerce").fillna(0.0)


def remap_simulated_features(
    simulated: pd.DataFrame,
    taxids: Sequence[str],
) -> pd.DataFrame:
    """Keep taxid rownames when present; otherwise map Feature* onto ranked taxids."""
    wanted = [str(t) for t in taxids]
    sim = simulated.copy()
    sim.index = sim.index.astype(str)
    overlap = [t for t in wanted if t in sim.index]
    if overlap and len(overlap) >= min(3, len(wanted)):
        out = sim.reindex(wanted).fillna(0.0)
        return out
    n = min(len(wanted), sim.shape[0])
    out = pd.DataFrame(0.0, index=wanted, columns=list(sim.columns))
    if n:
        out.iloc[:n, :] = sim.iloc[:n, :].to_numpy()
    return out


def simulate_count_matrix(
    observed: pd.DataFrame,
    *,
    mode: str = "sparsedossa2-fit",
    n_sample: int,
    config: Optional[Dict[str, Any]] = None,
) -> pd.DataFrame:
    cfg = apply_extra_flags(dict(config or {}))
    ranked = list(observed.sum(axis=1).sort_values(ascending=False).index.astype(str))
    template = cfg.get("sparsedossa2_template") or template_for_mode(mode)
    if cfg.get("sparsedossa2_fit"):
        template = None
    n_feature = int(cfg.get("n_feature") or cfg.get("sparsedossa2_n_feature") or max(len(ranked), 1))
    new_features = cfg.get("new_features")
    workers = _workers(cfg)
    seed = coerce_seed(cfg.get("seed", 42))
    verbose = bool(cfg.get("verbose") or cfg.get("sparsedossa2_verbose"))
    work = _job_workdir(cfg)
    tag = _job_tag("sim")
    inp = work / f"{tag}_observed.csv"
    outp = work / f"{tag}_simulated.csv"
    _write_feature_matrix(observed, inp)
    argv = [
        "simulate",
        "--output",
        str(outp),
        "--n-sample",
        str(int(n_sample)),
        "--n-feature",
        str(n_feature),
        "--seed",
        str(seed),
        "--workers",
        str(workers),
    ]
    if verbose:
        argv.append("--verbose")
    if template:
        argv.extend(["--mode", "template", "--template", str(template)])
        if new_features is None:
            argv.extend(["--new-features", "TRUE"])
    else:
        argv.extend(["--mode", "fit", "--fit", "--input", str(inp)])
        if new_features is None:
            argv.extend(["--new-features", "FALSE"])
    if new_features is not None:
        argv.extend(["--new-features", "TRUE" if new_features else "FALSE"])
    if cfg.get("fit_lambda") not in (None, ""):
        argv.extend(["--lambda", str(cfg["fit_lambda"])])
    _append_em_controls(argv, cfg)
    run_sparsedossa2_r(
        argv,
        timeout=int(cfg.get("timeout") or 7200),
        config=cfg,
    )
    simulated = _read_simulated_csv(outp)
    return remap_simulated_features(simulated, ranked)


def _fitcv_argv(
    inp: Path,
    outp: Path,
    cfg: Dict[str, Any],
) -> List[str]:
    folds = int(cfg.get("cv_folds") or cfg.get("K") or 5)
    workers = _workers(cfg)
    seed = coerce_seed(cfg.get("seed", 42))
    verbose = bool(cfg.get("verbose") or cfg.get("sparsedossa2_verbose"))
    lambdas = cfg.get("lambdas") or cfg.get("sparsedossa2_lambdas")
    argv = [
        "fitcv",
        "--input",
        str(inp),
        "--output",
        str(outp),
        "--cv-folds",
        str(max(2, min(folds, 100))),
        "--seed",
        str(seed),
        "--workers",
        str(workers),
    ]
    if verbose:
        argv.append("--verbose")
    if lambdas not in (None, ""):
        argv.extend(["--lambdas", str(lambdas)])
    _append_em_controls(argv, cfg)
    return argv


def _parse_cv_payload(path: Path) -> Dict[str, Any]:
    payload = json.loads(path.read_text(encoding="utf-8"))
    score = payload.get("cv_goodness_of_fit")
    try:
        score_f = float(score)
    except (TypeError, ValueError):
        score_f = float("nan")
    payload["cv_goodness_of_fit"] = score_f
    return payload


def concat_aligned_matrices(left: pd.DataFrame, right: pd.DataFrame) -> pd.DataFrame:
    """Feature-align two feature×sample tables and concatenate samples."""
    idx = sorted(set(left.index.astype(str)) | set(right.index.astype(str)))
    a = left.copy()
    b = right.copy()
    a.index = a.index.astype(str)
    b.index = b.index.astype(str)
    a = a.reindex(idx).fillna(0.0)
    b = b.reindex(idx).fillna(0.0)
    b.columns = [f"{c}__b" for c in b.columns]
    return pd.concat([a, b], axis=1)


def fitcv_score_matrix(
    matrix: pd.DataFrame,
    *,
    config: Optional[Dict[str, Any]] = None,
) -> Dict[str, Any]:
    rows = fitcv_score_jobs([("one", matrix)], config=config)
    return rows["one"]


def fitcv_score_jobs(
    items: Iterable[Tuple[str, pd.DataFrame]],
    *,
    config: Optional[Dict[str, Any]] = None,
) -> Dict[str, Dict[str, Any]]:
    require_sparsedossa2()
    cfg = apply_extra_flags(dict(config or {}))
    work = _job_workdir(cfg)
    timeout = int(cfg.get("timeout") or 7200)
    prepared: List[Tuple[str, Path, List[str]]] = []
    for key, matrix in items:
        tag = _job_tag("cv")
        inp = work / f"{tag}_{key}.csv"
        outp = work / f"{tag}_{key}.json"
        _write_feature_matrix(matrix, inp)
        argv = _fitcv_argv(inp, outp, cfg)
        prepared.append((str(key), outp, argv))

    out: Dict[str, Dict[str, Any]] = {}
    for key, outp, argv in prepared:
        try:
            run_sparsedossa2_r(argv, timeout=timeout, config=cfg)
            out[key] = _parse_cv_payload(outp)
        except Exception as exc:
            out[key] = {
                "cv_goodness_of_fit": float("nan"),
                "error": str(exc),
            }
    return out


def regenerate(
    annotation: Annotation,
    metadata: Optional[pd.DataFrame],
    config: Dict[str, Any],
) -> Dict[str, pd.DataFrame]:
    """``table_reads_generator`` entry point (fit or community template)."""
    _ = metadata
    cfg = apply_extra_flags(dict(config or {}))
    mode = str(
        cfg.get("table_reads_generator")
        or cfg.get("regeneration_mode")
        or "sparsedossa2-fit"
    )
    data = annotation.DataFrame
    if "sample" not in data.columns:
        data = data.copy()
        data["sample"] = "1"
    n_samples = _n_samples_or_observed(data, cfg.get("N"))
    n_reads = cfg.get("N_reads")
    if n_reads is not None:
        n_reads = int(n_reads)
    rescale = bool(cfg.get("rescale_abundance", False))
    threshold = float(cfg.get("threshold_amount", cfg.get("treshhold_amount", 1e-5)))
    result: Dict[str, pd.DataFrame] = {}
    for col in _taxid_columns(data):
        mat = _count_matrix(data, col, "sample")
        mat = _filter_taxa(mat, threshold, n_reads, rescale=False)
        if mat.empty:
            continue
        simulated = simulate_count_matrix(mat, mode=mode, n_sample=n_samples, config=cfg)
        simulated.columns = [str(c) for c in simulated.columns]
        simulated = _apply_abundance_scale(simulated, n_reads, rescale)
        name = _annotator_name(col)
        result[name] = _abundance_table_from_matrix(simulated, name)
    if not result:
        raise RuntimeError("SparseDOSSA2 produced no per-annotator abundance tables")
    return result


def register_sparsedossa2_tools(cfg: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
    """Write table generators + CV scorer into the install config."""
    from samovar.main_config import set_tool
    from samovar.paths import write_config

    config = dict(cfg or load_config())
    wrapper = str(Path(__file__).resolve())
    try:
        driver = str(driver_script().resolve())
    except MissingSparseDOSSA2Error:
        driver = str(Path(__file__).with_name("sparsedossa2.R"))
    rscript = shutil.which("Rscript") or ""
    try:
        rscript = rscript_executable()
    except MissingSparseDOSSA2Error:
        pass
    set_tool(
        config,
        "sparsedossa2.R",
        path=driver,
        env="",
        workflow="bash",
        group="workflow",
    )
    if rscript:
        set_tool(
            config,
            "Rscript",
            path=rscript,
            env="",
            workflow="bash",
            group="runtime",
        )
    table_flags = {
        "sparsedossa2-fit": "--fit",
        "sparsedossa2-stool": "--template Stool",
        "sparsedossa2-vaginal": "--template Vaginal",
        "sparsedossa2-ibd": "--template IBD",
    }
    for name, flags in table_flags.items():
        set_tool(
            config,
            name,
            path=wrapper,
            env="",
            workflow="bash",
            group="table_reads_generator",
            flags=flags,
        )
    set_tool(
        config,
        "sparsedossa2-cv",
        path=wrapper,
        env="",
        workflow="bash",
        group="scoring",
        flags="--cv-folds 5",
        inputs="regenerated/.regenerated_abundance",
    )
    write_config(config)
    return config
